scifin/statistics/statistics.py

```diff
@@ -132,27 +132,33 @@ def covariance_from_ts(list_ts, **kwargs):
     # Checks
     N = len(list_ts)
     T = list_ts[0].nvalues
+
     # Indexes must be identical
     idx = list_ts[0].data.index
     for i in range(1,N,1):
         if (list_ts[i].data.index != idx).all():
             raise AssertionError("Time series must have same index values.")
-    # Names must be different
+
+    # Forming a list of names
+    list_names = []
     set_names = set()
     for i in range(N):
+        list_names.append(list_ts[i].name)
         set_names.add(list_ts[i].name)
+
+    # Names must be different
     if len(set_names) != N:
         raise AssertionError("Names of time series must be different")
```

Comment:
not sure if I understand why it is necessary for the list of time series to be of different names here

Anyway, if you want to raise an Error when this happens, you can just catch it during the first loop above by something like
```
for i in range(N):
    time_series_name = list_ts[i].name
    if time_series_name in set_names:
        raise AssertionError("Names of time series must be different")

    set_names.add(list_ts[i].name)
```
In this case, the code will stop and raise the Error before looping through the whole list once there is a repeated names in the list found.

Also, you may consider define your own exception class in a separate module and import them here instead of AssertionError, so it is much more obvious what the error is about and easier to deal with them accordingly

```
class RepeatedTimeSeriesName(Exception):
    """
    Raise if the input list of TimeSeries contain members of identical names 

    """
    pass
```

Reply:
Very nice. I've implemented your lines of code, with one modification which is to keep the list of names rather than the set, because we actually want to keep the order.

The class is a good idea, but I don't think it is necessary now. This choice of different names among time series is a restriction that should not appear in many places. Here it is needed so that I can visualize with seaborn.heatmap the covariance matrix and have the names on axes.

Reply:
I guess we need to think about who is actually going to use this package. If the choice of different time-series names is only for visualisation and you to test the code, I don't feel it is justified to put the restriction here. 

Also, you can always create arbitrary names if needed (e.g. just naming it by the indices in the list).

Finally, refactor the codes to classes will make more object-oriented and much cleaner. And you can easily inherit  some the functions here for different algorithms/cases by create abstract/children classes (Feel like it would be better do the refactoring earlier so you will get used to it and have a clearer picture how different parts of the codes work together)

Reply:
Agreed. This restriction is not so useful in the end (I think of the simple case where we would have time series with None as a name, where this restriction is already problematic). So I've removed it.